src/Http/Http/src/Features/FormFeature.cs

```diff
@@ -181,27 +181,29 @@ private async Task<IFormCollection> InnerReadFormAsync(CancellationToken cancell
                         {
                             var fileSection = new FileMultipartSection(section, contentDisposition);
 
-                            // Enable buffering for the file if not already done for the full body
-                            section.EnableRewind(
-                                _request.HttpContext.Response.RegisterForDispose,
-                                _options.MemoryBufferThreshold, _options.MultipartBodyLengthLimit);
-
-                            // Find the end
-                            await section.Body.DrainAsync(cancellationToken);
-
                             var name = fileSection.Name;
                             var fileName = fileSection.FileName;
 
                             FormFile file;
                             if (section.BaseStreamOffset.HasValue)
                             {
+                                await section.Body.DrainAsync(cancellationToken);
+
                                 // Relative reference to buffered request body
                                 file = new FormFile(_request.Body, section.BaseStreamOffset.GetValueOrDefault(), section.Body.Length, name, fileName);
```

Comment:
This line would throw an exception if we were to use the default PipeReaderStream implementation.

Reply:
This part assumes it has a fully buffered seekable stream, that's why Length would be valid here. Hence we need to re-think how the buffering works.

Reply:
I understand what you are saying. 
As far as I know this will work, but I get that it doesn't align with the expected behavior of a non-seekable stream.
 
In that case I would appreciate it if you could help me figure this out. I guess I can think of something myself but it will take sometime since I am not as available anymore. 

The main issue as I see is how can we enable buffering into a PipeReader if PipeReader isn't seekable - should it be readable only once after buffering? Should we implement a mechanism that allows you to restart the PipeReader (possibly recreate it with resetting the underlying Stream)?  

The other issue is how to approach buffering in general - should we change it entirely to buffer from PipeReaders (BodyReader and MultipartPipeReader), or is it ok for now that we use the Streams similarly to how it works now. This is a smaller issue since it seems simple enough, but there might be some stuff that I am missing.

If the main problem is that length doesn't make sense in that context - I could just use a different DrainAsync functionality at the line above, and make it works so it also returns the length of the PipeReader.